File: lib/util/chatmsg.dart
import 'dart:io'; // file
import 'dart:convert'; // 用於 JSON 編碼
import 'package:intl/intl.dart';
import 'user.dart';

enum MessageType { text, image, file, textAndFile, system, request_news }
// enum MessageType { text, image, file, textAndFile, }

enum ServiceType {
  ai_reply,
  request_news,
  none,
  loginRegist,
  send_user_to_user,
  request_post,
  build_post,
}

// USER_TO_USER = "send_user_to_user"

MessageType whatMsgType(String text, File? image) {
  final hasText = text.isNotEmpty;
  final hasImage = image != null;

  if (hasText && hasImage) {
    return MessageType.textAndFile;
  } else if (hasImage) {
    return MessageType.file;
  } else {
    return MessageType.text;
  }
}

String GetNowTimeStamp() {
  final now = DateTime.now();
  String formatted = DateFormat('yyyy-MM-dd – kk:mm').format(now);
  // String formatted = DateTime.now().toIso8601String(); // 例如 2025-05-31T10:32:00.000

  return formatted;
}

DateTime ParseToDatetime(String timestr) {
  try {
    return DateFormat('yyyy-MM-dd – kk:mm').parse(timestr);
  } catch (e) {
    print('解析時間失敗：$e');
    return DateTime(2000); // fallback 時間，避免 crash
  }
}

String ChatMsg2String(ChatMsg msg) {
  return jsonEncode(msg.toJson());
}

class ChatMsg {
  final String sender;
  final String senderID;
  final String receiver;
  final String receiverID;

  // final TUser sender;
  // final TUser receiver;

  final MessageType type; // text / image / file 等
  final String content;
  final String timestamp;
  final ServiceType service;

  ChatMsg({
    required this.sender,
    this.senderID = '', // 預設為空字串
    required this.receiver,
    this.receiverID = '', // 預設為空字串
    this.service = ServiceType.none,
    this.type = MessageType.text,
    required this.content,
    required this.timestamp,
  });

  // ChatMsg({
  //   required this.sender,
  //   required this.receiver,
  //   this.service = ServiceType.none,
  //   this.type = MessageType.text,
  //   required this.content,
  //   required this.timestamp,
  // });

  // 轉成 JSON
  Map<String, dynamic> toJson() {
    return {
      'sender': sender,
      'senderID': senderID,
      'receiver': receiver,
      'receiverID': receiverID,
      'service': service.name,
      'type': type.name,
      'content': content,
      'timestamp': timestamp,
    };
  }

  factory ChatMsg.fromJson(Map<String, dynamic> json) {
    return ChatMsg(
      sender: json['sender'],
      senderID: json['senderID'] ?? '',
      receiver: json['receiver'],
      receiverID: json['receiverID'] ?? '',
      type: MessageType.values.firstWhere(
        (e) => e.name == json['type'],
        orElse: () => MessageType.text,
      ),
      content: json['content'],
      timestamp: json['timestamp'],
      service: ServiceType.values.firstWhere(
        (e) => e.name == json['service'],
        orElse: () => ServiceType.none,
      ),
    );
  }
}

File: Server/chatmsg.py
import json
from enum import Enum
from datetime import datetime


class MessageType(Enum):
    TEXT = "text"
    IMAGE = "image"
    FILE = "file"
    TEXT_AND_FILE = "textAndFile"
    SYSTEM = "system"
    REQ_NEWS = "request_news"


class ServiceType(Enum):
    loginRegist = "loginRegist"
    SEND_USER_TO_USER = "send_user_to_user"
    AI_REPLY = "ai_reply"
    REQ_NEWS = "request_news"
    UPDATE_USER = "update_user"
    REQ_POST = "request_post"
    BUILD_POST = "build_post"
    LOAD_USER = "load_user"
    NONE = "none"


def what_msg_type(text: str, image: str = None) -> MessageType:
    has_text = bool(text)
    has_image = image is not None

    if has_text and has_image:
        return MessageType.TEXT_AND_FILE
    elif has_image:
        return MessageType.FILE
    else:
        return MessageType.TEXT


def get_timestamp() -> str:
    now = datetime.now()
    return now.strftime("%Y-%m-%d – %H:%M")


class ChatMsg:
    def __init__(
        self,
        sender: str,
        receiver: str,
        content: str,
        timestamp: str,
        service: ServiceType = ServiceType.NONE,
        type: MessageType = MessageType.TEXT,
        senderID: str = "",
        receiverID: str = "",
    ):
        self.sender = sender
        self.senderID = senderID
        self.receiver = receiver
        self.receiverID = receiverID
        self.service = service
        self.type = type
        self.content = content
        self.timestamp = timestamp

    def to_json(self) -> dict:
        return {
            "sender": self.sender,
            "senderID": self.senderID,
            "receiver": self.receiver,
            "receiverID": self.receiverID,
            "service": self.service.value,
            "type": self.type.value,
            "content": self.content,
            "timestamp": self.timestamp,
        }

    @staticmethod
    def from_json(json_data: dict):
        return ChatMsg(
            sender=json_data.get("sender", ""),
            senderID=json_data.get("senderID", ""),
            receiver=json_data.get("receiver", ""),
            receiverID=json_data.get("receiverID", ""),
            service=ServiceType(json_data.get("service", "none")),
            type=MessageType(json_data.get("type", "text")),
            content=json_data.get("content", ""),
            timestamp=json_data.get("timestamp", ""),
        )


def chat_msg_to_string(msg: ChatMsg) -> str:
    return json.dumps(msg.to_json(), ensure_ascii=False)


# if __name__ == "__main__":

# msg = ChatMsg(
#     sender="Me",
#     receiver="TargetUser",
#     content="Hello",
#     timestamp=get_timestamp(),
#     type=what_msg_type("Hello"),
# )
# print(chat_msg_to_string(msg))
